pynamodb/attributes.py

```diff
@@ -325,3 +337,234 @@ def deserialize(self, value):
         Takes a UTC datetime string and returns a datetime object
         """
         return parse(value, dayfirst=False).datetime
+
+
+class MapAttributeMeta(type):
+    def __init__(cls, name, bases, attrs):
+        setattr(cls, '_attributes', None)
+
+
+class MapAttribute(with_metaclass(MapAttributeMeta, Attribute)):
+    attr_type = MAP
+
+    def __init__(self, **attrs):
+
+        hash_key = attrs.get('hash_key', False)
+        range_key = attrs.get('range_key', False)
+        null = attrs.get('null', None)
+        default = attrs.get('default', None)
+        attr_name = attrs.get('attr_name', None)
+
+        super(MapAttribute, self).__init__(hash_key=hash_key,
+                                           range_key=range_key,
+                                           null=null,
+                                           default=default,
+                                           attr_name=attr_name)
+        self.attribute_values = {}
+        self._set_attributes(**attrs)
+
+    def __iter__(self):
+        return iter(self.attribute_values)
+
+    def __getitem__(self, item):
+        return self.attribute_values[item]
+
+    @classmethod
+    def _get_attributes(cls):
+        """
+        Returns the list of attributes for this class
+        """
+        if cls._attributes is None:
+            cls._attributes = AttributeDict()
+            for item in [attr for attr in dir(cls) if
+                         isinstance(getattr(cls, attr), Attribute)]:
+                instance = getattr(cls, item)
+                if instance.attr_name is None:
+                    instance.attr_name = item
+                cls._attributes[item] = instance
+        return cls._attributes
+
+    def _set_attributes(self, **attrs):
+        """
+        Sets the attributes for this object
+        """
+        for attr_name, attr in self._get_attributes().aliased_attrs():
+            if attr.attr_name in attrs:
+                if type(attrs.get(attr.attr_name)) is type(attr) or not isinstance(attrs.get(attr_name), collections.Mapping):
```

Comment:
This function is probably the most complex and hardest to grasp. It would really appreciate some <3 with variable names and comments to show your intent

E.g what does `not isinstance(attrs.get(attr_name), collections.Mapping)` mean in terms of how we figure out where to set the value?